### PR TITLE
Add Reverse lookup

### PR DESCRIPTION
**Pull request by vincentdavis**
Issue #42, reverse lookup
I have implemented this as a class that initiates from the UUID dicts. I did it this way rather than writing the index to a file for two reasons; 1: it's fast enough, 2: In the future, we might want to add the ability to rebuild the index incases where the user adds custom UUIDs. 
Feedback appreciated.
I am not very familiar with taking full advantage of pytest.

### CODE REVIEW

File: src/bluetooth_numbers/reverse_lookup.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+import typing
+from typing import Literal
+from uuid import UUID
+
+from bluetooth_numbers import characteristic, company, descriptor, oui, service
+
+LOGIC = Literal["OR", "AND", "SUBSTR"]
+UUID_TYPES = Literal["characteristic", "company", "descriptor", "ouis", "service"]
+UUID_TYPE_DEFAULT = ("characteristic", "company", "descriptor", "ouis", "service")
```
Comment (koenvervloesem) [inline src/bluetooth_numbers/reverse_lookup.py:11]: The plural "ouis" seems inconsistent with the other types, which are singular.

Comment (vincentdavis) [reply src/bluetooth_numbers/reverse_lookup.py]: Agree, fixed

**Comment by vincentdavis** [comment]
Just made a few updates to fix doc test and test annotations

File: src/bluetooth_numbers/reverse_lookup.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+        Returns:
+            dict: Dictionary of terms to UUIDs.
+        """
+        reverse_lookup = {}
+        uuid_dicts = (
+            (characteristic, "characteristic"),
+            (company, "company"),
+            (descriptor, "descriptor"),
+            (oui, "ouis"),
```
Comment (koenvervloesem) [inline src/bluetooth_numbers/reverse_lookup.py:39]: Still a plural here.

File: src/bluetooth_numbers/reverse_lookup.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            logic: Search logic to use. Can be "OR", "AND" or "SUBSTR".
+
+        Returns:
+            set or Match, named tuples, (uuid, description, uuid_type)
+
+        Examples:
+            >>> from bluetooth_numbers.reverse_lookup import ReverseLookup, Match
+            >>> rl = ReverseLookup()
+            >>> matches = rl.lookup("Cycling Power")
+            >>> Match('00:05:5A', 'Power Dsine Ltd.', 'ouis') in matches
```
Comment (koenvervloesem) [inline src/bluetooth_numbers/reverse_lookup.py:70]: Also a plural "oui" here

**Comment by koenvervloesem** [comment]
Do you have pre-commit installed? It catches the same errors as found by the CI here, but locally.

See https://bluetooth-numbers.readthedocs.io/en/stable/contributing.html#clone-the-repository for the installation.

**Comment by vincentdavis** [comment]
> Do you have pre-commit installed? It catches the same errors as found by the CI here, but locally.
> 
> See https://bluetooth-numbers.readthedocs.io/en/stable/contributing.html#clone-the-repository for the installation.

I am working on that right now.

**Comment by vincentdavis** [comment]
@koenvervloesem Kinda struggling with the pre-commit. I have a lot of it cleaned up. 
I don't think I have everything installed to fully run the pre-commit and my editor, pycharm, is not using ruff configuration correctly. Its not giving me the hints and warnings or fixing the import sorting like I expect.


I am kinda stuck and out of time for now.

Maybe a development requirements file would be helpful?

**Comment by koenvervloesem** [comment]
Thanks already for the work you did! I'll take a look at the pre-commit errors tomorrow and see what I can fix. Maybe we can also ignore some of the errors if they're too pedantic.

File: src/bluetooth_numbers/reverse_lookup.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
+"""Reverse lookup class to find UUIDs by their description."""
+from __future__ import annotations
+
+from typing import Literal, NamedTuple, TYPE_CHECKING
```
Comment (koenvervloesem) [inline src/bluetooth_numbers/reverse_lookup.py:4]: `TYPE_CHECKING` needs to be first

File: src/bluetooth_numbers/reverse_lookup.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            logic="SUBSTR")
+            {Match(uuid=10853, description='Cycling Power Feature',
+            uuid_type='characteristic')}
+    """
+
+    def __init__(self) -> None:
+        """Initialize the ReverseLookup class, build index."""
+        self.index = self._build_index()
+
+    def _build_index(self) -> dict:
```
Comment (koenvervloesem) [inline src/bluetooth_numbers/reverse_lookup.py:53]: I think the return type can be specified as `dict[str, set[Match]]`.

File: src/bluetooth_numbers/reverse_lookup.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        """Build dictionary (index) of terms to UUIDs.
+
+        Returns:
+            dict: Dictionary of terms to UUIDs.
+        """
+        reverse_lookup = {}
```
Comment (koenvervloesem) [inline src/bluetooth_numbers/reverse_lookup.py:59]: The type can be specified as `dict[str, set[Match]]`

File: src/bluetooth_numbers/reverse_lookup.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        uuid_dicts = (
+            (characteristic, "characteristic"),
+            (company, "company"),
+            (descriptor, "descriptor"),
+            (oui, "oui"),
+            (service, "service"),
+        )
+        for uuid_dict, uuid_type in uuid_dicts:
+            for uuid, description in uuid_dict.items():
```
Comment (koenvervloesem) [inline src/bluetooth_numbers/reverse_lookup.py:68]: Let's add `# type: ignore[attr-defined]`, we don't need to make mypy 100% happy.

File: src/bluetooth_numbers/reverse_lookup.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+                for term in description.lower().split(" "):
+                    if term not in reverse_lookup:
+                        reverse_lookup[term] = set()
+                    reverse_lookup[term].add(Match(uuid, description, uuid_type))
+        return reverse_lookup
+
+    def lookup(
+        self,
+        terms: str,
+        uuid_types: list[UUID_TYPES] = UUID_TYPE_DEFAULT,
```
Comment (koenvervloesem) [inline src/bluetooth_numbers/reverse_lookup.py:78]: This does seem strange: the type is a list, the defaults are a tuple. If you're certain, we can add `# type: ignore[assignment]`, we don't need to make mypy 100% happy.

File: src/bluetooth_numbers/reverse_lookup.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        Args:
+            terms: String with the term(s) to search for.
+            uuid_types: List of UUID types to search in.
+            logic: Search logic to use. Can be "OR", "AND" or "SUBSTR".
+
+        Returns:
+            set or Match, named tuples, (uuid, description, uuid_type)
+        """
+        terms_set = set(terms.lower().split(" "))
+        results = set()
```
Comment (koenvervloesem) [inline src/bluetooth_numbers/reverse_lookup.py:92]: Let's add `# type: ignore[var-annotated]`, we don't need to make mypy 100% happy.

File: src/bluetooth_numbers/reverse_lookup.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+                        reverse_lookup[term] = set()
+                    reverse_lookup[term].add(Match(uuid, description, uuid_type))
+        return reverse_lookup
+
+    def lookup(
+        self,
+        terms: str,
+        uuid_types: list[UUID_TYPES] = UUID_TYPE_DEFAULT,
+        logic: LOGIC = "OR",
+    ) -> set:
```
Comment (koenvervloesem) [inline src/bluetooth_numbers/reverse_lookup.py:80]: Let's add `# type: ignore[type-arg]`, we don't need to make mypy 100% happy.

File: tests/test_reverse_lookup.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
+"""Test the bluetooth_numbers.reverse_lookup module."""
+import pytest
+
+from bluetooth_numbers.reverse_lookup import ReverseLookup, Match
```
Comment (koenvervloesem) [inline tests/test_reverse_lookup.py:4]: Swap `Match` and `ReverseLookup`.

**Comment by koenvervloesem** [comment]
I added some comments, these should solve/ignore most of the errors. But have a look at the doctests too.

**Comment by vincentdavis** [comment]
> I added some comments, these should solve/ignore most of the errors. But have a look at the doctests too.

Ok, I have cleaned up most of this, I'll finish up later today.

File: src/bluetooth_numbers/reverse_lookup.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+             Match(uuid=10851, description='Cycling Power Measurement',
+             uuid_type='characteristic'),
+             Match(uuid=10852, description='Cycling Power Vector',
+             uuid_type='characteristic'),
+             Match(uuid=10853, description='Cycling Power Feature',
+             uuid_type='characteristic'),
+             Match(uuid=10854, description='Cycling Power Control Point',
+             uuid_type='characteristic')}
+            >>> rl.lookup("Power Feature", uuid_types=['characteristic'],
+            logic="SUBSTR")
```
Comment (koenvervloesem) [inline src/bluetooth_numbers/reverse_lookup.py:49]: Make this `... logic="SUBSTR")  # doctest: +NORMALIZE_WHITESPACE` to make the doctest succeed.

File: src/bluetooth_numbers/reverse_lookup.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+            >>> rl.lookup("Cycling Power", logic="AND")
+            {Match(uuid=6168, description='Cycling Power', uuid_type='service'),
+             Match(uuid=10851, description='Cycling Power Measurement',
+             uuid_type='characteristic'),
+             Match(uuid=10852, description='Cycling Power Vector',
+             uuid_type='characteristic'),
+             Match(uuid=10853, description='Cycling Power Feature',
+             uuid_type='characteristic'),
+             Match(uuid=10854, description='Cycling Power Control Point',
+             uuid_type='characteristic')}
```
Comment (koenvervloesem) [inline src/bluetooth_numbers/reverse_lookup.py:47]: Sets don't have an ordering, so this output isn't guaranteed to be the same. You can see it by running `tox -e doctests`. Maybe better to illustrate this feature by doing a `Match(...) not in matches`?

**Comment by koenvervloesem** [comment]
Ok, just one issues with the import order (see above), and the two doctest issues I reviewed, and then this looks ready to merge.

**Comment by vincentdavis** [comment]
I think it's ready.
The `interrogate`  pre-commit does not work for me locally.  I like the idea of what it does. Maybe using Ruff's pydocstyle would be simpler and sufficient.
https://docs.astral.sh/ruff/rules/#pydocstyle-d

**Comment by koenvervloesem** [comment]
If you click on Details next to the failing tests, there are still some issues. Could you take a look at them? Once these tests run successfully, we can merge this PR.

**Comment by vincentdavis** [comment]
@koenvervloesem 
looking at the failing doc test, not sure what I am doing from.
```
Failed example:
    rl.lookup("Power Feature", uuid_types=['characteristic'],logic="SUBSTR")
Expected:
    {Match(uuid=10853, description='Cycling Power Feature',
    uuid_type='characteristic')}
Got:
    {Match(uuid=10853, description='Cycling Power Feature', uuid_type='characteristic')}
```

**Comment by koenvervloesem** [comment]
Yes, and the solution is the `# doctest: +NORMALIZE_WHITESPACE` I was talking about in https://github.com/koenvervloesem/bluetooth-numbers/pull/43#discussion_r1423619840

So this example should look like:

```pyhon
            >>> rl.lookup("Power Feature", uuid_types=['characteristic'],
            ... logic="SUBSTR")  # doctest: +NORMALIZE_WHITESPACE
            {Match(uuid=10853, description='Cycling Power Feature',
            uuid_type='characteristic')}

```

**Comment by vincentdavis** [comment]
Thanks for your patience

File: src/bluetooth_numbers/reverse_lookup.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+    Examples:
+            >>> from bluetooth_numbers.reverse_lookup import ReverseLookup, Match
+            >>> rl = ReverseLookup()
+            >>> matches = rl.lookup("Cycling Power")
+            >>> Match('00:05:5A', 'Power Dsine Ltd.', 'oui') in matches
+            True
+            >>> rl.lookup("Cycling Power Feature", logic="AND")
+            {Match(uuid=10853, description='Cycling Power Feature', uuid_type='characteristic')}
+            >>> rl.lookup("Power Feature", uuid_types=['characteristic'],logic="SUBSTR")
+            {Match(uuid=10853, description='Cycling Power Feature', uuid_type='characteristic')}
```
Comment (koenvervloesem) [inline src/bluetooth_numbers/reverse_lookup.py:41]: You can change this to the following to fix the last issue that Ruff finds:

```python
            >>> rl.lookup("Cycling Power Feature",
            ... logic="AND")  # doctest: +NORMALIZE_WHITESPACE
            {Match(uuid=10853, description='Cycling Power Feature',
            uuid_type='characteristic')}
            >>> rl.lookup("Power Feature", uuid_types=['characteristic'],
            ... logic="SUBSTR")  # doctest: +NORMALIZE_WHITESPACE
            {Match(uuid=10853, description='Cycling Power Feature',
            uuid_type='characteristic')}
```

**Comment by koenvervloesem** [comment]
Perfect! Now only the code coverage check complains, but the type checking line should be ignored. You can solve this by adding the following lines at the end of `.coveragerc`:

```

    # Assume if TYPE_CHECKING is covered
    if TYPE_CHECKING:
```

**Comment by koenvervloesem** [comment]
Thanks for patiently fixing your code to follow the project's code style!

**Pull request merged**